internal/servers/worker/internal/metric/proxy_ws_test.go

```diff
@@ -0,0 +1,371 @@
+package metric
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/require"
+)
+
+func TestInitializeWebsocketCollectors(t *testing.T) {
+	// Assert that calling the function with a nil Registerer is a valid operation.
+	require.NotPanics(t, func() { InitializeWebsocketCollectors(nil) })
+
+	// Try with an actual Registerer.
+	r := prometheus.NewRegistry()
+	require.NotPanics(t, func() { InitializeWebsocketCollectors(r) })
+
+	// Expect that some metrics exist.
+	f, err := r.Gather()
+	require.NoError(t, err)
+	require.Greater(t, len(f), 0)
```

Comment:
I notice you are using `require` exclusively in this file.  My understanding is that require should be used where progressing would cause a panic thus making it impossible to run subsequent tests.

`assert` should be used everywhere else since it is helpful to see all the assertions that are passing and failing with a single run if possible so the same test doesn't need to be run multiple times while fixes to assertions are made since `require` will stop the tests at the first require failure and thus make it harder to see all the other places assertions may be failing.

Reply:
I'm not totally sure what the pattern is in Boundary but in past PRs I've consistently only used `require` in tests. I suspect that so far this might be down to personal preference as I've seen `assert`-only test files, `require`-only test files, and anything in between.

Historically, I typically use `require` exclusively because when we write table tests the `t *testing.T` object inside the `t.Run` function is scoped exclusively for that particular test run. This means the entire overarching test does not abort if one test case fails - all test cases will still run (unless there's a panic somewhere of course).

For me, in general, using `assert` introduces more complexity to the tests (both to read and to write) as now I have to:
1. As the reader, understand why both `assert` and `require` were used and why they are used in the places they are;
2. As the writer, for each `assert` statement, understand how the overall test would continue in case the assertion was successful and unsuccessful. In cases where the assertion is complex (eg: asserting struct equality/struct value equality) this is not trivial. Many fields can be the cause for failure and I don't know which so I have to account for all in my following test code. Even worse, as assertions mount I have to keep in mind all the possible states from all the combinations of previous assertions.

With `require`, the test fails exactly where the assertion failed. Everything above is OK, and everything below is unknown. That's the only line I have to look at to start debugging process.

For me, while it is true that it does require that I run a particular test multiple times and this can be annoying (even for me!), I value the simplicity it affords more than any temporary annoyances that I might experience

I'm happy to discuss this further though if you think we could improve here! 😄

Reply:
FYI: I just learned about the problem with using require (ie. FailNow) in goroutines.  That isn't relevant here since there are no go routines, but it was something I wasn't aware of and thought I would share :).  See [this](https://github.com/hashicorp/boundary/pull/1978)